File: frontend/src/externals/pusher.js
import pusher from 'pusher-js';
import { ENVIRONMENT, PUSHER_APP_KEY, PUSHER_APP_CLUSTER } from '../const';

const PUSHER_ENABLED_ENVIRONMENTS = ["LIVE", "STAGE"]
let PUSHER_APP = null;

const getPusherApp = () => {
    console.log("PUSHER_APP: ", PUSHER_APP);
    if (PUSHER_ENABLED_ENVIRONMENTS.includes(ENVIRONMENT) && !PUSHER_APP){
        try{
            PUSHER_APP = new pusher(PUSHER_APP_KEY, {
                cluster: PUSHER_APP_CLUSTER,
                forceTLS: true,
            });
        }catch{
            console.log("Failed to create pusher app.")
            PUSHER_APP = null;
        }
    }
    console.log("PUSHER_APP: ", PUSHER_APP);
    return PUSHER_APP;
}

const subscribeChannel = (pusher_app, channel_name) => {
    if (!pusher_app){
        return null;
    }

    let channel = pusher_app.channel(channel_name);
    if(!channel){
        try{
            channel = pusher_app.subscribe(channel_name);
        }catch{
            console.log("Failed to subscribe channel.")
        }
    }

    return channel;
}


export {getPusherApp, subscribeChannel};

File: backend/src/externals/pusher.py
import logging
import traceback
import json
from django.conf import settings
from pusher import pusher
from decouple import config
from typing import Union, List, Optional
from helpers.custom_exception import convert_exception_string_to_one_line


logger = logging.getLogger("stdout")

PUSHER_APP_ID = config("PUSHER_APP_ID", default="PUSHER_APP_ID")
PUSHER_APP_KEY = config("PUSHER_APP_KEY", default="PUSHER_APP_KEY")
PUSHER_APP_SECRET = config("PUSHER_APP_SECRET", default="PUSHER_APP_SECRET")
PUSHER_APP_CLUSTER = config("PUSHER_APP_CLUSTER", default="PUSHER_APP_CLUSTER")

PUSHER_APP: Optional[pusher.Pusher] = None
PUSHER_ENABLED_ENVIRONMENTS = ["LIVE", "STAGE"]


def get_pusher_app():
    global PUSHER_APP
    if not PUSHER_APP and settings.ENVIRONMENT in PUSHER_ENABLED_ENVIRONMENTS:
        try:
            logger.info("Entering to create pusher app instance.")
            PUSHER_APP = pusher.Pusher(
                app_id=PUSHER_APP_ID,
                key=PUSHER_APP_KEY,
                secret=PUSHER_APP_SECRET,
                cluster=PUSHER_APP_CLUSTER,
                ssl=True,
            )
            logger.info("Pusher app created successfully.")
        except Exception:
            logger.warning("Failed to create pusher app.")
            PUSHER_APP = None

    logger.info("Created Pusher App: ", PUSHER_APP)
    return PUSHER_APP


def trigger_pusher(channels: Union[str, List[str]], event: str, data: dict) -> bool:
    if not channels:
        return False

    try:
        get_pusher_app()
        if not PUSHER_APP:
            return False
        PUSHER_APP.trigger(channels, event, data)
        logger.info("Pusher triggered successfully.")
        return True
    except Exception as e:
        logger.error(
            {
                "message": "Couldn't send data via pusher.",
                "channels": channels,
                "event": event,
                "data": {json.dumps(data)},
                "error": convert_exception_string_to_one_line(traceback.format_exc()),
            }
        )
        return False
